### PR TITLE
Check virtual memory size in memory mapping

### PR DESCRIPTION
**Pull request by tiger20081015**
Current RES_VASPACE_SIZE is defined as (CORE_MMU_PGDIR_SIZE * 10).
When requests virtual memory size larger than RES_VASPACE_SIZE,
there's no size check in core_mmu_add_mapping and the physic memory
out of RES_VASPACE_SIZE is mapped incorrectly.

For example, suppose CORE_MMU_PGDIR_SIZE = 1 << 21, i.e. RES_VASPACE_
SIZE = 20M bytes. When mmap table initilized, dump_mmap_table shows:

	E/TC:00 dump_mmap_table:704 type RES_VASPACE  va 0x112600000..0x1139fffff pa 0x00000000..0x013fffff size 0x01400000 (pgdir)
	E/TC:00 dump_mmap_table:704 type TA_RAM       va 0x113a00000..0x1153fffff pa 0x100600000..0x101ffffff size 0x01a00000 (pgdir)
	...
	E/TC:00 dump_mmap_table:704 type IO_SEC       va 0x117400000..0x117dfffff pa 0xc0000000..0xc09fffff size 0x00a00000 (pgdir)

after core_mmu_add_mapping(MEM_AREA_NSEC_SHM, paddr, 0x1a00000) executed:

	E/TC:00 dump_mmap_table:704 type RES_VASPACE  va 0x114000000..0x1139fffff pa 0x00000000..0xffffffffff9fffff size 0xffffffffffa00000 (pgdir)
	E/TC:00 dump_mmap_table:704 type TA_RAM       va 0x113a00000..0x1153fffff pa 0x100600000..0x101ffffff size 0x01a00000 (pgdir)
	...
	E/TC:00 dump_mmap_table:704 type IO_SEC       va 0x117400000..0x117dfffff pa 0xc0000000..0xc09fffff size 0x00a00000 (pgdir)
	E/TC:00 dump_mmap_table:704 type NSEC_SHM     va 0x112600000..0x113ffffff pa 0x102000000..0x1039fffff size 0x01a00000 (pgdir)

There are 2 mistakes in the lattest dumped mmu table:
1. For type RES_VASPACE, the base address 0x114000000 is even smaller
   than its end address 0x1139fffff;
2. For type TA_RAM and NSEC_SHM, the virtual memory range 0x113a00000
   ..0x113ffffff both exists, but the physics memory are different and
   virt_to_phys(0x113a00000) always returns 0x100600000.

This patch checks requested virtual memory size in core_mmu_add_mapping.
It also makes RES_VASPACE_SIZE configurable with macro
CFG_MMAP_RES_REGION_ENRTY_NUMS.

Signed-off-by: fangsuowu <fangsuowu@asrmicro.com>

### CODE REVIEW

File: core/arch/arm/mm/core_mmu.c [hunk 1]
```diff
@@ -52,7 +52,7 @@
 
 #include "core_mmu_private.h"
 
-#define RES_VASPACE_SIZE	(CORE_MMU_PGDIR_SIZE * 10)
+#define RES_VASPACE_SIZE	(CORE_MMU_PGDIR_SIZE * CFG_MMAP_RES_REGION_ENRTY_NUMS)
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu.c]: This is not related to the problem fixed below, please split into two patches.

Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu.c]: If va space shall be configurable, rather introduce a `CFG_RESERVED_VASPACE_SIZE`.

File: core/arch/arm/mm/core_mmu.c [hunk 2]
```diff
@@ -1465,6 +1465,9 @@ bool core_mmu_add_mapping(enum teecore_memtypes type, paddr_t addr, size_t len)
 	granule = 1 << tbl_info.shift;
 	p = ROUNDDOWN(addr, granule);
 	l = ROUNDUP(len + addr - p, granule);
+
+	if (map->size < l)
+		return false;
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c]: proposed comment `/* Ban overflowing virtual addresses */`

minor: can you add an empty line below? It was already missing.

File: core/arch/arm/mm/core_mmu.c (hunk 1 above)
Comment (tiger20081015) [reply core/arch/arm/mm/core_mmu.c]: new patches uploaded. thanks for reviewing.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`.  
@tiger20081015, I think you should correct your sign-off at least to get a capital letter to your real name. Please check the `Signed-off-by` bullet in [Contributing](https://github.com/OP-TEE/optee_os/blob/master/documentation/github.md#contributing-1) doc.

**Comment by jforissier** [comment]
@tiger20081015 please rebase, squash and apply the R-b tags.

**Comment by tiger20081015** [comment]
My patches doesn't touch below files. Below errors seems introduced by other guys?

> grub-core/osdep/linux/ofpath.c: In function ‘of_path_of_nvme’:
> grub-core/osdep/linux/ofpath.c:365:21: error: comparison between pointer and zero character constant [-Werror=pointer-compare]
>    if ((digit_string != '\0') && (*part_end == 'p'))
>                      ^~
> grub-core/osdep/linux/ofpath.c:365:8: note: did you mean to dereference the pointer?
>    if ((digit_string != '\0') && (*part_end == 'p'))
>         ^
> cc1: all warnings being treated as errors
> Makefile:9517: recipe for target 'grub-core/osdep/grub_mkrescue-ofpath.o' failed
> make[3]: *** [grub-core/osdep/grub_mkrescue-ofpath.o] Error 1
> make[3]: Leaving directory '/home/optee/devel/hikey/grub'
> Makefile:10906: recipe for target 'all-recursive' failed
> make[2]: *** [all-recursive] Error 1
> make[2]: Leaving directory '/home/optee/devel/hikey/grub'
> Makefile:3132: recipe for target 'all' failed
> make[1]: *** [all] Error 2
> make[1]: Leaving directory '/home/optee/devel/hikey/grub'
> bash: line 2: ./grub-mkimage: No such file or directory
> Makefile:299: recipe for target 'grub' failed
> make: *** [grub] Error 127
> make: *** Waiting for unfinished jobs....

**Comment by tiger20081015** [comment]
hi, is there any issues in my patch? when can my patches be merged?thanks.

**Comment by jforissier** [comment]
Hi @tiger20081015,
Your patches are fine. Thanks for the heads up -- it happens sometimes that patches are ready for merging with review tags applied etc. and I don't always notice. Sorry for the delay.
